### ISSUE TITLE
Confusing error when accessing a deeply nested field in dict with mixed types

### ISSUE THREAD

**Issue by marcofrasvda** [astral-sh/ty#2894]
### Question

Currently ty reports an error for an assignment that looks perfectly fine:
```py
payload = {
    "items": [
        {
            "uri": "https://example.com",
            "data": {
                "at": {
                    "time": "12:00",
                },
                "value": 5,
            },
        },
    ],
}

payload["items"][0]["data"]["value"] = 0
# Cannot assign to a subscript on an object of type `str` (invalid-assignment) [Ln 15, Col 1]
```

At first I though this was a bug as the error is not super straightforward here, it took me a minute to see why this happens, are there plans to support writing code like this (it's useful for tests) without having to create a specialized TypedDict? Or maybe make that error a bit more clear on exactly which part of the statement causes the error? Right now in VSCode and in the playground the whole statement is underlined which arguably made it harder to pinpoint the issue.

### Version

0.0.18

**Comment by sharkdp** [comment]
Thank you for reporting this.

I think there are two things that could be addressed here.

First, it's not completely unreasonable to expect a type checker to understand this code. In fact, this would already work if `payload["items"]` were a dict instead of a list. We would probably need to create narrowing places for int-subscripts (on lists) as well? (cc @ibraheemdev)

Second, the error message is very confusing, as you noted. The full diagnostic give a bit more context …
```
error[invalid-assignment]: Cannot assign to a subscript on an object of type `str`
  --> test.py:15:1
   |
13 | }
14 |
15 | payload["items"][0]["data"]["value"] = 0
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
16 | # Cannot assign to a subscript on an object of type `str` (invalid-assignment) [Ln 15, Col 1]
   |
info: The full type of the subscripted object is `Unknown | str | dict[Unknown | str, Unknown | dict[Unknown | str, Unknown | str] | int]`
info: `str` does not have a `__setitem__` method.
```
… but is still very difficult to understand. You are probably aware of this, but the problem is that `payload["items"][0]["data"]` is not narrowed to a `dict[…, …]` type, but to `str | dict[…, …]` (ignoring `Unknown` for the moment), where the `str` part comes in from the `"uri"` key, since we lost the ability to distinguish between `"data"` and `"uri"` when the 0-indexing happened. This is rather difficult to explain in an error message, but I'm sure we could do better here.

**Comment by AlexWaygood** [comment]
> First, it's not completely unreasonable to expect a type checker to understand this code. In fact, this would already work if `payload["items"]` were a dict instead of a list. We would probably need to create narrowing places for int-subscripts (on lists) as well? (cc [@ibraheemdev](https://github.com/ibraheemdev))

We attempted this in https://github.com/astral-sh/ruff/pull/22929, but it led to a fairly significant performance regression with an underwhelming ecosystem report, so we decided not to go down that route for now

**Comment by carljm** [comment]
I don't think this is a compelling case for narrowing int subscripts on lists, since it looks like the list is in fact intended to be homogenous (every element the same type), so it would be wasteful to narrow each int subscript independently. If the list became long, that would become very expensive.

This case rather just points out a limitation of place-based dict-key narrowing -- it fails when a dict is nested inside a type which we don't support place narrowing for. This is a case where inferring an implicit synthesized `TypedDict` type could work better than place narrowing.

Maybe at some point we try to improve this implicit narrowing, but I think for now defining an explicit `TypedDict` type for each item is the best option.

Regarding the diagnostic, I think perhaps the most obvious improvement would be for the LHS annotation to more specifically show exactly which subscript fails, rather than highlighting the entire LHS

**Comment by AlexWaygood** [comment]
> since it looks like the list is in fact intended to be heterogenous 

You mean homogenous ;)

**Comment by carljm** [comment]
> You mean homogenous ;)

Er, yes :) Edited, thanks.
